contracts/gateway/GatewayBase.sol

```diff
@@ -0,0 +1,395 @@
+pragma solidity ^0.4.23;
+
+import './GatewayLib.sol';
+import "./CoreInterface.sol";
+import "./SafeMath.sol";
+import "./MessageBus.sol";
+
+/**
+ *  @title GatewayBase contract.
+ *
+ *  @notice GatewayBase contains general purpose functions shared between
+ *  gateway and co-gateway contract.
+ */
+contract GatewayBase {
+
+    using SafeMath for uint256;
+    /** Emitted whenever a Gateway/CoGateway contract is proven.
+     *	wasAlreadyProved parameter differentiates between first call and replay
+     *  call of proveGateway method for same block height
+     */
+    event GatewayProven(
+        address _gateway,
+        uint256 _blockHeight,
+        bytes32 _storageRoot,
+        bool _wasAlreadyProved
+    );
+
+
+    /* Struct */
+
+    /**
+     * ActiveProcess stores the information related to in progress process
+     * like stake/mint unstake/redeem.
+     */
+    struct ActiveProcess {
+
+        /** latest message hash. */
+        bytes32 messageHash;
+
+        /** Outbox or Inbox process. */
+        MessageBus.MessageBoxType messageBoxType;
+    }
+
+
+    bytes32 constant STAKE_TYPEHASH = keccak256(
+        abi.encode(
+            "Stake(uint256 amount,address beneficiary,MessageBus.Message message)"
+        )
+    );
+
+    bytes32 constant REDEEM_TYPEHASH = keccak256(
+        abi.encode(
+            "Redeem(uint256 amount,address beneficiary,MessageBus.Message message)"
+        )
+    );
+    bytes32 constant GATEWAY_LINK_TYPEHASH = keccak256(
+        abi.encode(
+            "GatewayLink(bytes32 messageHash,MessageBus.Message message)"
```

Comment:
Add `Message` struct as specified in EIP-712 for `STAKE_TYPEHASH`, `REDEEM_TYPEHASH `, `GATEWAY_LINK_TYPEHASH `

Reply:
Created issue for this.
#347